CONTRIBUTING: enhance section about squashing and rebasing (#22256)
RIOT-OS/RIOT: CONTRIBUTING.md

`fixup` or `DONOTMERGE`. These commits have to be squashed or removed before a
merge is permitted.

***Watch out:*** Don't squash your commits unless a maintainer asks you to do it.

Otherwise the history of review changes is lost and for large PRs, it
makes it difficult for the reviewer to follow them. It might also happen that
you introduce a regression and won't be able to recover your working state from
previous commits.

Squashing a commit is done using the `rebase` subcommand of git in interactive
mode, you can find information on rebasing in the
[GitHub rebase documentation][about-git-rebase].

[about-git-rebase]: https://help.github.com/articles/about-git-rebase/

The easiest method is to use the following command that compares the changes
in your branch to the `master` branch and shows you all the commits you have
made since in a text editor. You can select, drop, reorder and squash the
commits here.

```console
$ git rebase master -i
```

***Note:*** This will automatically rebase your branch to the `master` branch
afterwards. If a lot of development happened on the `master` branch since you
have opened your Pull Request, a merge conflict might arise.

> 

Reviewer: Not sure whether I like that advice, I'd normally suggest going to the last commit before your own commits on your own branch, rebasing on master while squash has caused a lot of issues for me in the past, I'd rather `git rebase -i <PRE_CHANGES COMMIT>` and then `git rebase master`

Reviewer 2: I agree here. Resolving merge conflict due to reordering/squashing while also resolving merge conflicts at the same time can be confusing. Doing the squashing first (and fix any merge conflicts due that) and then reparent the branch on top of `master` does split the two merge conflict stories and IMO makes it more manageable.

Author: To clarify: `git rebase master -i` was in the tutorial before I edited it, I only learned about that existing by reading/editing it 🤣 

But yes, generally that is my workflow too. Squash first, then rebase if necessary.

Reviewer: You touch it you fix it :nerd_face: :point_up: 

(You dont have to its fine)

Author: Well of course I'll fix it, but I don't have a definite idea about what to write.

Author: I removed that section and explained the commit-hash-based method.

While directly rebasing on `master` can be wanted, it can also lead to unwanted side effects. Someone experienced won't read the contribution guide anyways, much less the parts about how to use git.

If you used [fixup commits](#add-fixup-commits-during-review) during the review
phase, squashing commits can be performed in a single command:

```console
$ git rebase -i --autosquash master
```

If you want to remain on your branch and avoid possible merge conflicts at
this stage, you can set the starting point relative to the last commit, which
is called `HEAD` in git terminology:

```console
$ git rebase -i HEAD~n
```

You have to replace the `n` with the number of commits you want to go back in
history.
You can find out how many commits you made by listing commits with the `log`
subcommand. For example to show the last 10 commits, you can execute:

```console
$ git log --oneline -n10

> ```

Reviewer: same goes with HEAD-n, I feel like simply specifying the exact commit hash before the changes works out better for most cases, you can even use the git history tab on github if you don't want to deal with searching through a git log

Reviewer 2: I personally tent to do both. I'm not 100% sure were to draw the line between explaining git here, or just refer to other existing doc for further tips and advanced uses. After all, using git isn't that RIOT specific.

Author: That is very true. Do you happen to have any good git guides on hand that we could link to?

I personally found that one of the biggest challenges when I started and the man pages are confusing AF 😅

Author: I shortened the explanation. While I also tend to use `HEAD~n`, it's likely that beginners will struggle with that.

My "personal goal" here is to have a guide at hand which we can point to when typical git problems arise in the contribution process.

### Squash commits after review

During the course of a Pull Request, a lot of `fixup!` commits can accumulate.
To keep the project commit history clean, these commits have to be combined
(squashed) into sensible commits. This is a good opportunity to take a look

> at again our [commit conventions].

Reviewer 3: ```suggestion
(squashed) into sensible commits. This is a good opportunity to take a look
again at our [commit conventions].
```

***Note:*** The static tests will warn you about no-merge keywords such as
`fixup` or `DONOTMERGE`. These commits have to be squashed or removed before a
merge is permitted.

***Watch out:*** Don't squash your commits unless a maintainer asks you to do it.

Otherwise the history of review changes is lost and for large PRs, it
makes it difficult for the reviewer to follow them. It might also happen that
you introduce a regression and won't be able to recover your working state from
previous commits.

Squashing a commit is done using the `rebase` subcommand of git in interactive
mode, you can find information on rebasing in the
[GitHub rebase documentation][about-git-rebase].

[about-git-rebase]: https://help.github.com/articles/about-git-rebase/

The safest method to squash your commits is to find out the last commit hash

> of the `master` branch with `git log --oneline -n10`. Depending on the number of

Reviewer 3: ```suggestion
of the `master` branch with `git merge-base HEAD master`. Depending on the number of
```

Reviewer 3: and remove the sentence behind it.

Author: Wow, I'm even learning new tricks here :D

commits you made, `-n` can be bigger or smaller of course.

You can then execute `git rebase -i <last master hash>`, which will show you
all the commits you have made since in a text editor. You can select, drop,
reorder and squash the commits here.

If you used [fixup commits](#add-fixup-commits-during-review) during the review
phase, squashing commits can be performed in a single command:

```console
$ git rebase -i --autosquash <last master hash>
```

If you encounter a merge conflict, it is generally easiest to use a merge tool
like [meld](https://meldmerge.org/) or the built-in merge tool of your editor
or IDE.

After the merge conflict is resolved you can continue the rebase by using

```console
$ git rebase --continue
```

Once squashing is done, you will have to force push your branch to update the
PR:

```console
$ git push origin <your branch> --force-with-lease
```

### Manually rebasing

> Sometimes a maintainer will ask you to rebase your Pull Request for various

Reviewer 3: ```suggestion
A maintainer might ask you to rebase your Pull Request for various
```
